### DISCUSSION TITLE
What is the best way to create targets that are similar to the result of a for loop?

### DISCUSSION THREAD

**Discussion by kendonB** [ropensci/targets#615] [Help]
I have a use case (a stochastic dynamic program) that involves a loop that depends on the output of the previous stage of the loop - a classic for loop use case for R.

What's the best way to do this in targets such that each stage is a target? Is there a tarchetype for this use case?

Each stage of the loop is a parallel computation as well (looping over the state variable nodes) which ideally would be parallelised using targets' `map`.

**Comment by wlandau** [comment]
There is not a built-in target factory for this, but static branching can be recursive. Here is an example pipeline that runs 3 parallel Fibonacci sequences.

``` r
library(targets)
tar_script({
  library(rlang)
  library(tarchetypes)
  f <- function(x_n_1, x_n_2) {
    x_n_1 + x_n_2
  }
  values <- list(
    n = seq(3, 10),
    x_n_1 = syms(sprintf("x_%s", seq(2, 9))),
    x_n_2 = syms(sprintf("x_%s", seq(1, 8)))
  )
  list(
    tar_target(x_1, rep(0, 3)),
    tar_target(x_2, rep(1, 3)),
    tar_map(
      values = values,
      names = n,
      tar_target(x, f(x_n_1, x_n_2), pattern = map(x_n_1, x_n_2))
    )
  )
})

tar_manifest()
#> # A tibble: 10 × 3
#>    name  command     pattern      
#>    <chr> <chr>       <chr>        
#>  1 x_1   rep(0, 3)   <NA>         
#>  2 x_2   rep(1, 3)   <NA>         
#>  3 x_3   f(x_2, x_1) map(x_2, x_1)
#>  4 x_4   f(x_3, x_2) map(x_3, x_2)
#>  5 x_5   f(x_4, x_3) map(x_4, x_3)
#>  6 x_6   f(x_5, x_4) map(x_5, x_4)
#>  7 x_7   f(x_6, x_5) map(x_6, x_5)
#>  8 x_8   f(x_7, x_6) map(x_7, x_6)
#>  9 x_9   f(x_8, x_7) map(x_8, x_7)
#> 10 x_10  f(x_9, x_8) map(x_9, x_8)
```

<sup>Created on 2021-08-26 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>
